lib/Dataflow_analysis.ml

```diff
@@ -0,0 +1,800 @@
+   variable dependencies of the "y" variable, printing results to stdout.
+*)
+let analysis_example (mir : stmt_loc prog) : dataflow_graph =
+  let var_table, model_block = mir.modelb in
+  let df_graph = block_dataflow_graph model_block var_table in
+  let var = "y" in
+  let label_deps = final_var_dependencies df_graph true (Var var) in
+  let expr_deps = preexisting_var_dependencies df_graph label_deps in
+  let data_vars = ExprSet.of_list [Var "x"] in
+  let prior_term_labels =
+    exprset_independent_target_terms df_graph data_vars
+  in
+  let prior_terms =
+    List.map (LabelSet.to_list prior_term_labels) ~f:(fun l ->
+        match (LabelMap.find_exn df_graph.node_info_map l).loc with
+        | TargetTerm {term; _} -> term
+        | _ -> raise (Failure "Found non-target term in target term list") )
+  in
+  let preexisting_vars =
+    ExprSet.of_list
+      (List.map ("target" :: Map.Poly.keys var_table) ~f:(fun v -> Var v))
+  in
+  if false then (
+    Sexp.pp_hum Format.std_formatter
+      [%sexp (df_graph.node_info_map : node_info_fixpoint LabelMap.t)] ;
+    print_string "\n\n" ;
+    print_endline
+      ( "Preexisting variables: "
+      ^ Sexp.to_string [%sexp (preexisting_vars : ExprSet.t)] ) ;
+    print_endline
+      ( "Target term nodes: "
+      ^ Sexp.to_string [%sexp (df_graph.target_term_nodes : LabelSet.t)] ) ;
+    print_endline
+      ( "Possible endpoints: "
+      ^ Sexp.to_string [%sexp (df_graph.possible_exits : LabelSet.t)] ) ;
+    print_endline
+      ( "Assumed data variables: "
+      ^ Sexp.to_string [%sexp (data_vars : ExprSet.t)] ) ;
+    print_endline
+      ( "Data-independent target term expressions: "
+      ^ Sexp.to_string [%sexp (prior_terms : expr list)] ) ;
+    print_endline
+      ( "Var " ^ var ^ " depends on labels: "
+      ^ Sexp.to_string [%sexp (label_deps : LabelSet.t)] ) ;
+    print_endline
+      ( "Var " ^ var ^ " depends on preexisting variables: "
+      ^ Sexp.to_string [%sexp (expr_deps : ExprSet.t)] ) ) ;
+  df_graph
+
+(***********************************)
+(* Tests                           *)
+(***********************************)
+
+(* Inline tests don't seem to work *)
+
+(* let%test _ = 5 = 120
+   let%test _ = raise (Failure "ran test")*)
+
+let%expect_test "Example program" =
+  let ast =
+    Parse.parse_string Parser.Incremental.program
+      "      model {\n\
+      \              for (i in 1:2)\n\
+      \                for (j in 3:4)\n\
+      \                  print(\"Badger\");\n\
+      \            }\n\
+      \            "
+  in
+  let mir =
+    Ast_to_Mir.trans_prog "" (Semantic_check.semantic_check_program ast)
+  in
+  let table, block = mir.modelb in
+  let df_graph = block_dataflow_graph block table in
+  print_s [%sexp (df_graph : dataflow_graph)] ;
+  [%expect
+    {|
+      ((node_info_map
+        ((0
+          ((dep_sets (() (((Var target) 0)))) (possible_previous ()) (rhs_set ())
```

Comment:
Just because it's used kind of like a pre-existing global variable. I don't think it makes much difference to the algorithm.

Reply:
I was imagining we'd only add it in there if its value was used, right? And I don't think we allow its value to be used anymore (need to double-check that). Might be a better indication of e.g. a function tampering with the value.

Reply:
If it couldn't be used as RHS variable, I could probably simplify a bit of the code, and in that case I'd probably argue for a change in the MIR to support target increments as its own case rather than an Assignment.